stdlib/parser/dsl.md

```diff
@@ -0,0 +1,197 @@
+# Introduction
+
+This file gives a basic runthrough of the in-progress design of a DSL
+for syntax and AST specification.
+
+## Basics
+
+```
+-- Line comments
+
+/-
+Multiline comments
+-/
+```
+
+We declare new types in the AST and non-terminals in
+the grammar with `type`.
+
+```
+type Expr
+```
+
+A constructor in the AST and production in the grammar is introduced
+with `prod`.
+
+```
+prod TmInt: Expr = value:Int
+```
+
+Each constructor gets its own language fragment in the generated code,
+thus the above yields the following fragment:
+
+```
+lang TmInt
+  syn Expr =
+  | TmInt { value : Int }
+```
+
+The syntax of a single production can contain regex style repetition
+operators:
+
+```
+-- Repeated field names are collected into a sequence in the
+-- order they appear in the source.
+prod TmTuple: Expr = "(" fields:Expr ("," fields:Expr)+ ")"
+-- Zero-or-more is expressed with `?` and produce an Option
+prod TmLet: Expr = "let" ident:LIdent (":" tyIdent:Type)? "=" value:Expr "in" body:Expr
+```
+We can also give the constructor nested records:
+
+```
+prod TmMatch: Expr = "match" target:Expr "with" arms:{"|" pat:Pat "->" body:Expr}+
+```
+
+```
+lang TmMatch
+  syn Expr =
+  | TmMatch { target : Expr, arms : [{pat : Pat, body : Expr}]}
+end
+```
+
+Explicit grouping (i.e. parentheses) should not be specified as a
+production but rather explicitly as grouping:
+
+```
+grouping "(" Expr ")"
+```
+
+It should have the form `grouping` `<token>` `<type>` `<token>`.
+
+### Tokens
+
+Tokens can be either a literal (e.g. `"let"`) or a named token
+(e.g. `LIdent`). In the current system there is a number of builtin
+token types, currently specified in `lexer.mc`.
+
+## Operators
+
+Operators can be written either explicitly:
+```
+prod TmAdd: Expr = left:Expr "+" right:Expr
+```
+
+...or with syntactic sugar:
+```
+infix TmMul: Expr = "*"
+
+-- Similarly for prefix and postfix
+prefix TmNot: Expr = "!"
+postfix TmFieldAccess: Expr = "." field:LIdent
+```
+
+Each production will be one of these things:
+- Simple/atomic if it has no direct recursion on either edge
+- Prefix if it only has direct recursion on the right edge
+- Postfix if it only has direct recursion on the left edge
+- Infix if it has direct recursion on both edges
+
+I suspect that productions that sometimes have direct recursion on the
+edge and sometimes not are a bad idea. Please let me know if such
+cases come up, so we can see if that is correct or not.
```

Comment:
I don't understand how this can happen. Do you have an example of such a production?